service/src/main/java/com/codeforcommunity/processor/ReservationProcessorImpl.java

```diff
@@ -0,0 +1,206 @@
+    reservationsRecord.setBlockId(completeReservationRequest.getBlockID());
+    reservationsRecord.setUserId(userData.getUserId());
+    reservationsRecord.setTeamId(completeReservationRequest.getTeamID());
+    reservationsRecord.setActionType("complete");
+    reservationsRecord.setPerformedAt(new Timestamp(System.currentTimeMillis()));
+
+    Optional<ReservationsRecord> maybeReservation = lastAction(reservationsRecord.getBlockId());
+
+    // check if block is reserved by the user (already checked if user is on team)
+    if (maybeReservation.isPresent()) {
+      if (!(maybeReservation.get().getUserId().equals(userData.getUserId()))) {
+        throw new BlockNotReservedException(reservationsRecord.getBlockId());
+      }
+    } else {
+      throw new BlockNotReservedException(reservationsRecord.getBlockId());
+    }
+
+    reservationsRecord.store();
+  }
+
+  @Override
+  public void releaseReservation(
+      JWTData userData, ReleaseReservationRequest releaseReservationRequest) {
+    basicChecks(releaseReservationRequest.getBlockID(), userData.getUserId(), null);
+
+    ReservationsRecord reservationsRecord = db.newRecord(RESERVATIONS);
+    reservationsRecord.setBlockId(releaseReservationRequest.getBlockID());
+    reservationsRecord.setUserId(userData.getUserId());
+    reservationsRecord.setActionType("release");
+    reservationsRecord.setPerformedAt(new Timestamp(System.currentTimeMillis()));
+
+    Optional<ReservationsRecord> maybeReservation = lastAction(reservationsRecord.getBlockId());
+
+    // check if block is reserved by the user (already checked if user is on team)
+    if (maybeReservation.isPresent()) {
+      if (!(maybeReservation.get().getUserId().equals(userData.getUserId()))) {
+        throw new BlockNotReservedException(reservationsRecord.getBlockId());
+      }
+    } else {
+      throw new BlockNotReservedException(reservationsRecord.getBlockId());
+    }
+
+    reservationsRecord.store();
+  }
+
+  @Override
+  public void uncompleteReservation(
+      JWTData userData, UncompleteReservationRequest uncompleteReservationRequest) {
+    basicChecks(uncompleteReservationRequest.getBlockID(), userData.getUserId(), null);
+
+    ReservationsRecord reservationsRecord = db.newRecord(RESERVATIONS);
+    reservationsRecord.setBlockId(uncompleteReservationRequest.getBlockID());
+    reservationsRecord.setUserId(userData.getUserId());
+    reservationsRecord.setActionType("uncomplete");
+    reservationsRecord.setPerformedAt(new Timestamp(System.currentTimeMillis()));
+
+    Optional<ReservationsRecord> maybeReservation = lastAction(reservationsRecord.getBlockId());
+
+    // check if the user is an admin and if the block has been completed
+    if (userData.getPrivilegeLevel().equals(PrivilegeLevel.STANDARD)) {
+      throw new AuthException("User does not have the required privilege level.");
+    }
+
+    if (maybeReservation.isPresent()) {
+      if (!(maybeReservation.get().getActionType().equals("complete"))) {
```

Comment:
Why are we comparing strings instead of enums here?

Reply:
Enum wasn't behaving as expected and forgot to fix, but now resolved